### DISCUSSION TITLE
Admin panel application.

### DISCUSSION THREAD

**Discussion by BartMommens** [spatie/laravel-multitenancy#245] [General]
Hello all,

Currently i'm having trouble trying to figure out how to do the Admin panel for a Multi-tenancy (multi db) Saas Application.

So i'm using Laravel Fortify and Sanctum for all my tenants apps, but the downside of this package is that it's only able to use one guard being 'web' guard. All tenants have shared roles and permissions ( stored in Landlord database). 

So currenly i'm thinking about creating a different laravel project to manage the landlord database. In this App i'll be using a CRUD for tenant roles / permissions.  This App may on a separate domain with it's own users (landlord admins) and so on...

But i'm not sure if this is a good approach, to have a separate application to manage the landlord database.  It will have a few 'shared' models between the projects such as roles and permissions, and should be able to use the same app key as the multi tenant application.

Is it a good approach to create a separate application or not at all? I think i would build a separate laravel package to keep the models synced. My biggest problem is actually that i'm afraid of doing something that isn't concerned as a good practice. And since most guides, tutorials and docs suggest running the Admin app inside the same project i'm facing a dilemma.

The biggest advange of the Separate admin application (my opinion) would be that i could run a copy of my multi tenant system on a separate server (for example a clients server's with his own tenants) and then i can just switch landlord database in the admin panel and manage those apps separatly.

What do you think of my idea? Maybe this topic can also help other users facing this dilemma 😊

Thanks for your feedback, any feedback is welcome.

Looking forward to your replies.

Best regards,
Bart

**Comment by shinanmhd** [comment]
Hi @BartMommens 
I am faced with the same problem.
Could you please let me know what solution you implemented?
thank you

**Reply by BartMommens** [reply]
Hey @shinanmhd,

Currently I parked this 'problem'. Since the lack of time I just add and configure tenants with tinker. The only downside is that everything is done via command line. Over the months / year I've bundled most commands so it takes almost no effort.

Building a separate admin application would mean you need to keep 2 applications up to date, and since it would have a minimal usage it would be more work maintaining admin application than actually doing the work via tinker.

So far I don't regret the decision to focus on the main app.  Tinker is a pretty handy tool.

Sorry I can't help you out further on this topic.

Kind regards,
Bart

**Reply by BENRAHAL-Aicha** [reply]
@shinanmhd  Hi , did you find any solution ?

**Comment by FranciscoDumont** [comment]
I’m also running into this situation, would love to know if there’s any recommended approach.

**Reply by BartMommens** [reply]
Hey @FranciscoDumont,

As mentioned above, i parked this approach since there was no time / budget to work this out.
All the manipulations of the Landlord database and tenant settings are being done via commandline. If we had build 2 applications one ADMIN application to control the "Client" application it would mean that we needed to keep 2 applications up to date as Laravel kept updating. It wasn't deemed cost efficient and we just have a playbook with custom commandline scripts that we customize for each new tenant that we onboard. The entire process takes about 30-45 minutes, from setting up the domain, configuring the DNS records and deploying the tenant).  So we abandoned the Admin APP since there are only 2 people responsible for the onboarding of a new tenant.

So unfortunately i haven't solved this approach entirely

**Reply by FranciscoDumont** [reply]
Thanks a lot for sharing your experience, it makes complete sense. 
I’ll keep an eye on this thread in case someone comes up with another solution in the future.

**Reply by BartMommens** [reply]
Sorry I couldn't be of more help here. Have a nice weekend and good luck with the project

**Reply by BENRAHAL-Aicha** [reply]
@FranciscoDumont hey , did u manage to find any solution , i'm facing same issue

**Comment by MattWiseParking** [comment]
Hi Bart,
Both approaches are legitimate, but I'd push back gently on the premise that's driving you toward a separate app: Fortify being "web guard only" isn't really true, and it's solvable inside one project.
The guard problem is smaller than it looks. Fortify's guard is configurable (config/fortify.php → guard), but more importantly, you don't need Fortify for the landlord side at all. You can define a second guard in config/auth.php — e.g. an admin guard with its own provider pointing at an Admin model whose $connection is your landlord database — and write a plain login controller for it (it's ~20 lines with Auth::guard('admin')->attempt()). Fortify keeps handling your tenant users on web; the admin guard lives alongside it. Spatie Permission also supports multiple guards, so landlord roles/permissions can be scoped to the admin guard while tenant ones stay on web.
Routing-wise, Spatie Multitenancy already anticipates this. The tenant-aware middleware (NeedsTenant, EnsureValidTenantSession) is only applied to the routes you choose. You can have a landlord.yourapp.com (or /admin) route group with no tenant middleware, running purely on the landlord connection. One codebase, two "faces".
On the separate-app approach: it works, but the costs are real and ongoing. Shared models via a package means every schema change now touches three repos (package, tenant app, admin app), you have to keep migrations in sync, and you'll hit awkward questions like "which app owns the migration for the roles table?" Sharing the APP_KEY across apps is also a smell — you'd only need that if you're sharing encrypted payloads or sessions, and if you are, that coupling suggests they want to be one app anyway.
Your portability argument (pointing one admin panel at different landlord databases on different servers) is genuinely interesting, but note you can get the same result from a single codebase: deploy the same app to the client's server with a different .env, and the landlord admin comes along for free. If you truly need one central panel managing multiple landlord databases remotely, that's better solved with an API — expose landlord management endpoints from each installation and build a thin dashboard that talks to them — rather than direct cross-server database connections, which get painful fast (firewalling, latency, credential management, no code-level validation).
My suggestion: start with the single app + landlord route group + dedicated admin guard. It's the well-trodden path, it's why most guides suggest it, and it keeps your options open. If you later genuinely need the multi-installation management story, extract an API then — extraction is much easier than premature separation.
Best of luck with it!

**Comment by MattWiseParking** [comment]
when you spin up a new db instance, i would also run the migrate and seeder, that way you can have a seeder file that populates the roles and permissions.  Keep each tenant db to have their own roles and permissions, cross refrencing to the landlord to obtain roles and permissions is a headache to program, keep database separate from each other, that way each tenant could in theory have different functionality enabled/disabled.

**Reply by MattWiseParking** [reply]
be specific to your migrate and running seeder, ensure you specify the tenant ID otherwise you would wipe all data against the other live databases

**Comment by shinanmhd** [comment]
Sorry, I missed the message.

What I’ve done in several projects using Spatie Multitenancy is keep the landlord admin and tenant apps in the same Laravel codebase while separating them through dedicated domains, guards, database connections, and middleware.

In the project I was working on when I originally asked this question, I used Spatie’s `SwitchTenantDatabaseTask` with separate `landlord` and `tenant` connections.

In the application I’m currently working on which also uses Spatie Multitenancy, I use a shared-database approach with tenant-scoped data and permissions. The platform admin still has its own domain, authentication guard, user model, routes, and authorization boundary.

I’d only build a separate admin application if it needed to manage multiple independently deployed installations. In that case, I’d use a secure management API—not direct database switching, shared models, or a shared `APP_KEY`.
